src/preprocessing/passes/bv_ackermann.cpp

```diff
@@ -0,0 +1,173 @@
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Ackermannization preprocessing pass.
+ **
+ ** This implements the Ackermannization preprocessing pass, which enables
+ ** very limited theory combination support for eager bit-blasting via
+ ** Ackermannization. It reduces constraints over the combination of the
+ ** theories of fixed-size bit-vectors and uninterpreted functions as
+ ** described in
+ **   Liana Hadarean, An Efficient and Trustworthy Theory Solver for
+ **   Bit-vectors in Satisfiability Modulo Theories.
+￼**   https://cs.nyu.edu/media/publications/hadarean_liana.pdf
+ **/
+
+#include "preprocessing/passes/bv_ackermann.h"
+
+#include "expr/node.h"
+#include "options/bv_options.h"
+#include "theory/bv/theory_bv_utils.h"
+
+#include <unordered_set>
+
+using namespace CVC4;
+using namespace CVC4::theory;
+
+namespace CVC4 {
+namespace preprocessing {
+namespace passes {
+
+/* -------------------------------------------------------------------------- */
+
+namespace
+{
+
+void storeFunction(
+    TNode func,
+    TNode term,
+    FunctionToArgsMap& fun_to_args,
+    SubstitutionMap& fun_to_skolem)
+{
+  if (fun_to_args.find(func) == fun_to_args.end())
+  {
+    fun_to_args.insert(make_pair(func, NodeSet()));
+  }
+  NodeSet& set = fun_to_args[func];
+  if (set.find(term) == set.end())
+  {
+    set.insert(term);
+    Node skolem = bv::utils::mkVar(bv::utils::getSize(term));
+    fun_to_skolem.addSubstitution(term, skolem);
+  }
+}
+
+void collectFunctionSymbols(
+    TNode term,
+    FunctionToArgsMap& fun_to_args,
+    SubstitutionMap& fun_to_skolem,
+    std::unordered_set<TNode, TNodeHashFunction>& seen)
+{
+  if (seen.find(term) != seen.end()) return;
+  if (term.getKind() == kind::APPLY_UF)
+  {
+    storeFunction(term.getOperator(), term, fun_to_args, fun_to_skolem);
+  }
+  else if (term.getKind() == kind::SELECT)
+  {
+    storeFunction(term[0], term, fun_to_args, fun_to_skolem);
+  }
+  else
+  {
+    AlwaysAssert(term.getKind() != kind::STORE,
+                 "Cannot use eager bitblasting on QF_ABV formula with stores");
+  }
+  for (unsigned i = 0; i < term.getNumChildren(); ++i)
+  {
+    collectFunctionSymbols(term[i], fun_to_args, fun_to_skolem, seen);
+  }
+  seen.insert(term);
+}
+
+}  // namespace
+
+/* -------------------------------------------------------------------------- */
+
+BVAckermann::BVAckermann(PreprocessingPassContext* preprocContext)
+    : PreprocessingPass(preprocContext, "bv-ackermann"),
+      d_funcToSkolem(preprocContext->getUserContext())
+{
+}
+
+PreprocessingPassResult BVAckermann::applyInternal(
+    AssertionPipeline* assertionsToPreprocess)
+{
+  Assert(options::bitblastMode() == theory::bv::BITBLAST_MODE_EAGER);
+  AlwaysAssert(!options::incrementalSolving());
+
+  std::vector<Node> assertions(assertionsToPreprocess->ref());
```

Comment:
Don't use the assertions vector directly. Assertions need to be replaced via `assertionsToPreprocess.replace(...)` since this ensures that all dependencies for proof generation are tracked correctly.